### ISSUE TITLE
Create a text layout mechanism for interacting with large-ish bodies of text.

### ISSUE THREAD

**Issue by schell** [schell/odin#3]
When working with larger bodies of text (like > 4 words), measuring the bounds of text gets really expensive. You can see here how slow the prog gets while editing text.

![thu-aug-18-09-57-01-pdt-2016](https://cloud.githubusercontent.com/assets/24942/17782884/6687e5dc-652a-11e6-93af-74e801f06ff9.gif)

I think I could write a layout "engine" that only re-renders and re-measures each new word, and saves each processed word in a map keyed by the word itself.

Attached are the prof and ps files from a similar run.

[Thu-Aug-18-09:48:52-PDT-2016 copy.txt](https://github.com/schell/odin/files/425465/Thu-Aug-18-09.48.52-PDT-2016.copy.txt)
[Thu-Aug-18-09:48:52-PDT-2016 copy.pdf](https://github.com/schell/odin/files/425466/Thu-Aug-18-09.48.52-PDT-2016.copy.pdf)

**Comment by schell** [comment]
@Peaker - sorry to pull you in on my project, but you probably have experience with rendering lots of text with lamdu. Have you run into this problem before? What are you using to render and layout text in lamdu?

**Comment by schell** [comment]
A little more spelunking and it looks like it's FTGL via [graphics-drawingcombinators](https://hackage.haskell.org/package/graphics-drawingcombinators-1.5.1/docs/Graphics-DrawingCombinators.html#g:6), added in https://github.com/lamdu/lamdu/commit/f9aa3a403aacfad9c967cec945662ad576cf0b78

**Comment by Peaker** [comment]
Font rendering was far harder than I anticipated.

Text has no width, really, but a bounding box and an "advance".

We switched from FTGL which doesn't have subpixel rendering to freetype-gl with some minor modifications, that does. It's also a c codebase around libfreetype so easier to compile everywhere.

It does the width computations for us and I've not noticed slowness of text geometry computations, but I've not done extensive profiling either.
